---
title: "Summary of progressive selection of variants Run3"
author: "Kiranmayee Bakshy"
date: "June 21, 2018"
output: html_document
---

**A summary of the filtration_run3 is as follows**
 **Run3 consists of variants with a) homozygous GT in target animals b) no heterozygous genotype found in non-target animals c) all INFO filters other than DP**
 
  * Kept 2093896 out of 14397424 variant sites
  * Removed 1664776 singletons
  * Removed 342817 sites that were heterozygous in all other regions
  * Removed 406107 sites with AN < 337
  * Removed 229484 sites with MQ0F >= 0.1
  * Removed 4819784 sites with MQSB <= 0.95
  * Removed 132708 sites with 328> DP >9002

 **Summary statistics for INFO fields** 
  
```{r echo=FALSE}
plot_info<-readRDS("filtered_run3/filtered_run3.vcf.gz.rds")
summary(plot_info)
```

**Plink summary statistics such as MAF, AAF, fraction missing per sample, fraction missing per variant, HWE**
   
```{r echo=FALSE}
#Summary of filtered dataset MAF
df<-readRDS("filtered_run3/filtered_run3.frq.rds")
summary(df)

#Summary of filtered dataset AAF
library(data.table)
df<-fread("filtered_run3/filtered_run3.afreq", header=T, select=c("ID", "ALT_FREQS"))
summary(df)

#Summary of filtered dataset fraction missing per variant
df<-readRDS("filtered_run3/filtered_run3.lmiss.rds")
summary(df)

#Summary of filtered dataset fraction missing per sample
df<-read.delim("filtered_run3/filtered_run3.imiss", header=T, sep="")
summary(df[,c(2,6)])
```
